drivers/video/video_mcux_mipi_csi2rx.c

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright 2024 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nxp_mipi_csi2rx
+
+#include <fsl_mipi_csi2rx.h>
+
+#include <zephyr/drivers/video.h>
+#include <zephyr/kernel.h>
+
+#define LOG_LEVEL CONFIG_LOG_DEFAULT_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(mipi_csi);
+
+#define DEFAULT_MIPI_CSI_NUM_LANES 2
+#define DEFAULT_CAMERA_FRAME_RATE  30
+
+struct mipi_csi2rx_config {
+	const MIPI_CSI2RX_Type *base;
+	const struct device *sensor_dev;
+};
+
+struct mipi_csi2rx_data {
+	csi2rx_config_t csi2rxConfig;
+};
+
+static int mipi_csi2rx_set_fmt(const struct device *dev, enum video_endpoint_id ep,
+			       struct video_format *fmt)
+{
+	const struct mipi_csi2rx_config *config = dev->config;
+	struct mipi_csi2rx_data *drv_data = dev->data;
+	csi2rx_config_t csi2rxConfig = {0};
+	uint8_t i = 0;
+
+	/*
+	 * Initialize the MIPI CSI2
+	 *
+	 * From D-PHY specification, the T-HSSETTLE should in the range of 85ns+6*UI to 145ns+10*UI
+	 * UI is Unit Interval, equal to the duration of any HS state on the Clock Lane
+	 *
+	 * T-HSSETTLE = csi2rxConfig.tHsSettle_EscClk * (Tperiod of RxClkInEsc)
+	 *
+	 * csi2rxConfig.tHsSettle_EscClk setting for camera:
+	 *
+	 *    Resolution  |  frame rate  |  T_HS_SETTLE
+	 *  =============================================
+	 *     720P       |     30       |     0x12
+	 *  ---------------------------------------------
+	 *     720P       |     15       |     0x17
+	 *  ---------------------------------------------
+	 *      VGA       |     30       |     0x1F
+	 *  ---------------------------------------------
+	 *      VGA       |     15       |     0x24
+	 *  ---------------------------------------------
+	 *     QVGA       |     30       |     0x1F
+	 *  ---------------------------------------------
+	 *     QVGA       |     15       |     0x24
+	 *  ---------------------------------------------
+	 */
+	static const uint32_t csi2rxHsSettle[][4] = {
+		{
+			1280,
+			720,
+			30,
+			0x12,
+		},
+		{
+			1280,
+			720,
+			15,
+			0x17,
+		},
+		{
+			640,
+			480,
+			30,
+			0x1F,
+		},
+		{
+			640,
+			480,
+			15,
+			0x24,
+		},
+		{
+			320,
+			240,
+			30,
+			0x1F,
+		},
+		{
+			320,
+			240,
+			15,
+			0x24,
+		},
+	};
+
+	csi2rxConfig.laneNum = DEFAULT_MIPI_CSI_NUM_LANES;
```

Comment:
We should probably make this configurable with devicetree, right? For MIPI-DSI drivers we have a discrete interface that supports attaching and detaching (which we might want for MIPI-CSI drivers in the future as well). For now though, there should still be a way to change this without modifying the source code IMO.

Reply:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/34665051/81796442-0113-457d-ba8e-3a9537233906)

It seems the MIPI CSI2 supports 1 or 2 data lanes. While I agree there is probably work to be done extending the video API with endpoints, why wouldn't we expose this right now as a configurable property? This feels like something a customer will run into pretty soon with this driver if we lock them to using 2 lanes when the IP supports 1 data lane

Reply:
Sorry, it is miswritten in my previous comment. I mean the ov5640 driver does not support 1 data lane (not the MIPI CSI2Rx). There is currently no camera driver that supports 1 data lanes. So, I think we should not expose this in the devicetree (even as a temporary solution) as it is never tested on any hardware.

In the future, when there is a camera driver supporting it, we could do so as your suggestion (again as a temporary solution). But **the data lanes need to be declared in both MIPI CSI2RX node and the camera sensor node** (and not only the number of lanes, the order of the lanes does matter also).

Another reason, 1-data lane camera sensor is rare as it is not efficient as 2 or 4 data lanes.